designs/v1-api.md

```diff
@@ -0,0 +1,529 @@
+3. Users update their `v1beta1` manifests that they are applying through IaC or GitOps to use the new `v1` version.
+4. Karpenter drops the `v1beta1` version from the `CustomResourceDefinition` and the conversion webhooks on the next minor version release of Karpenter, leaving no webhooks and only the `v1` version still present
+
+## NodePool API
+
+```
+apiVersion: karpenter.sh/v1
+kind: NodePool
+metadata:
+  name: default
+spec:
+  template:
+    metadata:
+      labels:
+        billing-team: my-team
+      annotations:
+        example.com/owner: "my-team"
+    spec:
+      nodeClassRef:
+        group: karpenter.k8s.aws # Updated since only a single version will be served
+        kind: EC2NodeClass
+        name: default
+      taints:
+        - key: example.com/special-taint
+          effect: NoSchedule
+      startupTaints:
+        - key: example.com/another-taint
+          effect: NoSchedule
+      requirements:
+        - key: "karpenter.k8s.aws/instance-category"
+          operator: In
+          values: ["c", "m", "r"]
+          minValues: 2 # Alpha field, added for spot best practices support
+  disruption:
+    budgets:
+      - nodes: 0
+        schedule: "0 9 * * mon-fri"
+        duration: 8h
+      - nodes: "10%"
+    expireAfter: 720h | Never
+    consolidationPolicy: WhenUnderutilized | WhenEmpty
+    consolidateAfter: 1m | Never # Added to allow additional control over consolidation aggressiveness
+  weight: 10
+  limits:
+    cpu: "1000"
+    memory: 1000Gi
+status:
+  conditions:
+    # Abnormal True Conditions
+    - lastTransitionTime: "2024-02-02T19:54:34Z"
+      status: "True"
+      type: NodeClassReady
+    - lastTransitionTime: "2024-02-02T19:54:34Z"
+      status: "True"
+      type: Ready
+  resources:
+    cpu: "2"
+    memory: "100Mi"
+    ephemeral-storage: "100Gi"
+```
+
+### Defaults
+
+Defaults are unchanged in the v1 API version but are included in the API proposal for completeness.
+
+`disruption.budgets` : `{"nodes": "10%"}`
+`disruption.expireAfter`: `30d/720h`
+`disruption.consolidationPolicy`: `WhenUnderutilized`
+`disruption.consolidateAfter`: `15s`
+
+### Printer Columns
+
+**Category:** Stability, Breaking
+
+#### Current
+
+```
+➜  karpenter git:(main) ✗ kubectl get nodepools -o wide
+NAME      NODECLASS   WEIGHT
+default   default     100
+fallback  fallback    
+```
+
+#### Proposed
+
+```
+➜  karpenter git:(main) ✗ kubectl get nodepools -o wide
+NAME      NODECLASS  NODES  READY  AGE   WEIGHT  CPU%  MEMORY%
+default   default    4      True   2d7h  100     5.8%  99%
+fallback  default    100    True   2d7h          20%   20%
+```
+
+**Standard Columns**
+
+1. Name
+2. NodeClass - Allows users to easily see the NodeClass that the NodePool is using
```

Comment:
There's a problem: we don't know what the kind is. With this API, there could be two different node classes with the same name but different kinds.

Reply:
How big of a problem do you think that is in the printer columns? I'm hesitant to want to include both the name and the kind/group in an API when most users may name them differently for clarity anyways. Perhaps the wide columns, but even there I'm not fully convinced that it's needed at this point since no CloudProvider supports multiple kinds for NodeClasses right now.

Reply:
Agree with Jonathan on this one. I don't know if anyone is mixing providers